lib/assert.js

```diff
@@ -1001,6 +1001,38 @@ assert.ifError = function ifError(err) {
   }
 };
 
+/**
+ * @param {never} actual
+ * @param {string | Error} [message]
+ */
+assert.unreachable = function unreachable(actual, message) {
+  const argsLen = arguments.length;
+  if (argsLen < 1) {
+    throw new ERR_MISSING_ARGS('actual');
+  }
+
+  let internalMessage = false;
+  if (argsLen === 1) {
+    message = 'Expected this assertion to be unreachable';
+    internalMessage = true;
+  }
+
+  if (message instanceof Error) throw message;
+
+  const err = new AssertionError({
```

Comment:
I think we should be using `innerFail`
```suggestion
  innerFail({
```

Reply:
I don't understand innerFail. It's not used in fail and it throws obj.message when it is an error, which isn't documented anywhere.

```
assert.fail({ message: new Error() })
// Throws the error
```

The instance of Error check you highlighted if different than the check in inner fail.

Reply:
Accepting that change caused tests to fail because generatedMessage is not true when the default message is used.

```
> new assert.AssertionError({message: "foo", generatedMessage: true }).generatedMessage
false
```

I can't pass generatedMessage to the AssertionEerror constructor. What do you think about adding a generatedMessage parameter to innerFail? Or a `defaultMessage` and have innerFail work out generated and message?

Reply:
I was thinking we could add another branch for `unreachable`, as it's done with https://github.com/nodejs/node/blob/ab8942848b9967e6e0289ff7f1a8b12a349b7a28/lib/internal/assert/assertion_error.js#L362-L363
It's also possible that `innerFail` is not the correct util for what we're trying to do, I haven't looked that closely to the code.